crates/iota-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -284,6 +284,131 @@ module std::vector {
         }
     }
 
+    public macro fun select_do_ref<$T>($v: &vector<$T>, $ix: &vector<u64>, $f: |u64, &$T|) {
+        let v = $v;
+        let ix = $ix;
+        let v_len = v.length();
+        ix.do_ref!(|i| {
+            assert!(*i < v_len); // can't access EINDEX_OUT_OF_BOUNDS outside of std::vector
+            $f(*i, &v[*i]);
+        });
+    }
+
+    public macro fun select_do_mut<$T>($v: &mut vector<$T>, $ix: &vector<u64>, $f: |u64, &mut $T|) {
+        let v = $v;
+        let ix = $ix;
+        let v_len = v.length();
+        ix.do_ref!(|i| {
+            assert!(*i < v_len);
+            $f(*i, &mut v[*i]);
+        });
+    }
+
+    public macro fun select_do_with_ix_ref<$T>($v: &vector<$T>, $ix: &vector<u64>, $f: |u64, u64, &$T|) {
+        let v = $v;
+        let ix = $ix;
+        let v_len = v.length();
+        ix.length().do!(|k| {
+            let i = ix[k];
+            assert!(i < v_len);
+            $f(k, i, &v[i]);
+        });
+    }
+
+    public macro fun select_do_with_ix_mut<$T>($v: &mut vector<$T>, $ix: &vector<u64>, $f: |u64, u64, &mut $T|) {
+        let v = $v;
+        let ix = $ix;
+        let v_len = v.length();
+        ix.length().do!(|k| {
+            let i = ix[k];
+            assert!(i < v_len);
+            $f(k, i, &mut v[i]);
+        });
+    }
+
+    public macro fun select_find_index<$T>($v: &vector<$T>, $ix: &vector<u64>, $f: |&$T| -> bool): Option<u64> {
+        'select_find_index: {
+            select_do_ref!($v, $ix, |i,x| {
+                if ($f(x)) return 'select_find_index option::some(i);
+            });
+            option::none()
+        }
+    }
+
+    public macro fun select_map_ref<$T, $U>($v: &vector<$T>, $ix: &vector<u64>, $f: |u64,&$T| -> $U): vector<$U> {
+        let mut u = vector::empty<$U>();
+        select_do_ref!($v, $ix, |i,x| u.push_back($f(i,x)));
+        u
+    }
+
+    public macro fun select_collect<$T>($v: &vector<$T>, $ix: &vector<u64>): vector<$T> {
+        select_map_ref!($v, $ix, |_,x| *x)
+    }
+
+    public macro fun select_top_n<$T>($v: &vector<$T>, $n: u64, $less_than: |&$T, &$T| -> bool): vector<u64> {
```

Comment:
I'm not sure if it's a good idea to have with such specific methods. Something more generic like `zip/unzip` and `sort` (or numpy-like `argsort`) would be more generic and could achieve the same thing.